lib/ABI/ModelAlignment.cpp

```diff
@@ -0,0 +1,171 @@
+
+namespace abi {
+
+static RecursiveCoroutine<std::optional<uint64_t>>
+naturalAlignmentImpl(model::VerifyHelper &VH,
+                     const model::Type &Type,
+                     model::ABI::Values ABI);
+static RecursiveCoroutine<std::optional<uint64_t>>
+naturalAlignmentImpl(model::VerifyHelper &VH,
+                     const model::QualifiedType &Type,
+                     model::ABI::Values ABI);
+
+template<typename FieldType>
+static RecursiveCoroutine<std::optional<uint64_t>>
+aggregateAlignmentImpl(model::VerifyHelper &VH,
+                       const SortedVector<FieldType> Fields,
+                       model::ABI::Values ABI) {
+  uint64_t Alignment = 0;
+
+  for (const auto &Field : Fields) {
+    if (auto A = rc_recur naturalAlignmentImpl(VH, Field.Type, ABI))
+      Alignment = std::max(Alignment, *A);
+    else
+      rc_return std::nullopt;
+  }
+
+  rc_return Alignment;
+}
+
+static RecursiveCoroutine<std::optional<uint64_t>>
+naturalAlignmentImpl(model::VerifyHelper &VH,
+                     const model::Type &Type,
+                     model::ABI::Values ABI) {
+  auto MaybeAlignment = VH.size(&Type);
+  if (MaybeAlignment)
+    rc_return MaybeAlignment;
+
+  // This code assumes that the type `Type` is well formed.
+  uint64_t Alignment;
+
+  switch (Type.Kind) {
+  case model::TypeKind::RawFunctionType:
+  case model::TypeKind::CABIFunctionType:
+    // Function prototypes have no size - hence no alignment.
+    Alignment = 0;
+    break;
+
+  case model::TypeKind::PrimitiveType:
+    // The alignment of primitives is simple to figure out based on the abi
+    if (const auto *P = llvm::cast<model::PrimitiveType>(&Type);
+        P->PrimitiveKind == model::PrimitiveTypeKind::Void) {
+      // Make sure void types have no size.
+      if (P->Size != 0)
+        rc_return std::nullopt;
+
+      Alignment = 0;
+    } else {
+      Alignment = abi::getScalarStackAlignment(ABI)[P->Size].AlignAt;
+    }
+    break;
+
+  case model::TypeKind::EnumType: {
+    // The alignment of an enum is the same as the alignment of its underlying
+    // type
+    const auto &Underlying = llvm::cast<model::EnumType>(&Type)->UnderlyingType;
+    if (auto A = rc_recur naturalAlignmentImpl(VH, Underlying, ABI))
+      Alignment = *A;
+    else
+      rc_return std::nullopt;
+  } break;
+
+  case model::TypeKind::TypedefType: {
+    // The alignment of an enum is the same as the alignment of its underlying
+    // type
+    const auto &Typedef = llvm::cast<model::TypedefType>(&Type)->UnderlyingType;
+    if (auto A = rc_recur naturalAlignmentImpl(VH, Typedef, ABI))
+      Alignment = *A;
+    else
+      rc_return std::nullopt;
+  } break;
+
+  case model::TypeKind::StructType: {
+    // The alignment of a struct is the same as the alignment of its most
+    // strictly aligned member.
+    const auto &Fields = llvm::cast<model::StructType>(&Type)->Fields;
+    if (auto A = rc_recur aggregateAlignmentImpl(VH, Fields, ABI))
```

Comment:
Why `Impl` suffix?

Reply:
Ahah, why not just `aggregateAlignement`? Do you already have such a function?

Reply:
I had it in an earlier design, but that got cut. Only the `internal helper` remained. Either way, I'll rename it.

Reply:
(keeping this open as a reminder about `Impl` suffixes. Will take a look closer to the merge point)

Reply:
Closing this as I've just got rid of the last two `Impl`s (*):
`distributeImpl()` -> `distributeOne()`.
`needsVectorRegisterSupportImpl()` -> `needsVectorRegisterSupport()`

* in the code introduced by this PR.